x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/LimitedRole.java

```diff
@@ -27,17 +27,13 @@
  * provided role.
  */
 public final class LimitedRole extends Role {
-    private final Role limitedBy;
+    private final Role fromRole;
 
     LimitedRole(String[] names, ClusterPermission cluster, IndicesPermission indices, ApplicationPermission application,
-            RunAsPermission runAs, Role limitedBy) {
+            RunAsPermission runAs, Role fromRole) {
```

Comment:
This refactoring is necessary because the `LimitedRole#names` (which is what's used by the audit) reported the names of the role descriptors from the API Key definition. It now returns the names of the owner user roles.

Reply:
If the intention is to have `LimitedRole#names` to report creator's roles instead of the sythetic roles. I wonder whether it would be better to just override `Role#names` instead of changing it here. This change works as well, but it makes the code hard to read unless you go all the way to change all relevant names:
* [indicesAccessControl.limitIndicesAccessControl(...)](https://github.com/elastic/elasticsearch/pull/58928/files#diff-b6f74647fa674cfdc4fc6505f3acac27L72)
* [limitedByIndicesAccessControl](https://github.com/elastic/elasticsearch/pull/58928/files#diff-b6f74647fa674cfdc4fc6505f3acac27R65)
* [resourcePrivilegesMapForLimitedRole](https://github.com/elastic/elasticsearch/pull/58928/files#diff-b6f74647fa674cfdc4fc6505f3acac27R116)
* Even the class name itself `LimitedRole` kinda suggests the field of name `limitedBy`.

We also may wanna change the order of the conditional checks, e.g.:
```
return super.checkClusterAction(action, request, authentication) && fromRole.checkClusterAction(action, request, authentication);
```
It might be better to check against `fromRole` first, since common cases would be `fromRole` having less privileges than the `limitedBy` role.

But I wonder whether all above changes are worthwhile if the goal is just to get the role names of the creator. It might be simpler to just override the `names` method unless there is some concern that I am missing. Also, this is a remote concern, but if we decide to support derived keys by using multiple `limitedBy` roles, this change would make it harder to do so.

Reply:
> If the intention is to have LimitedRole#names to report creator's roles instead of the synthetic roles

This is the intention indeed.

> I wonder whether it would be better to just override Role#names instead of changing it here.

That was my first intention as well, but here's how the [limited role creation](https://github.com/elastic/elasticsearch/blob/0c2b689b8c69ccf5d2d96267e903f877178a7a23/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/LimitedRole.java#L188) looked like:

```
    public static LimitedRole createLimitedRole(Role fromRole, Role limitedByRole) {
        Objects.requireNonNull(limitedByRole, "limited by role is required to create limited role");
        return new LimitedRole(limitedByRole.names(), fromRole.cluster(), fromRole.indices(), fromRole.application(),
                fromRole.runAs(), limitedByRole);
    }
```

After I raised the PR I realised I need to invert the conjunctions as you've described and maybe forgot to push that.

Overriding `LimitedRole#names` is another option I can make that work.

Reply:
@ywangd I went with your suggestion of:

> It might be simpler to just override the names method unless there is some concern that I am missing

In a seeparate PR https://github.com/elastic/elasticsearch/pull/59041 .

Reply:
Thanks @albertzaharovits I commented on the other PR.